crates/ruff_server/src/lint.rs

```diff
@@ -24,17 +26,20 @@ use crate::{edit::ToRangeExt, PositionEncoding, DIAGNOSTIC_NAME};
 #[derive(Serialize, Deserialize, Debug, Clone)]
 pub(crate) struct AssociatedDiagnosticData {
     pub(crate) kind: DiagnosticKind,
-    pub(crate) fix: Fix,
+    pub(crate) fix: Option<Fix>,
     pub(crate) code: String,
+    pub(crate) noqa_edit: Option<ruff_diagnostics::Edit>,
 }
 
-/// Describes a fix for `fixed_diagnostic` that applies `document_edits` to the source.
+/// Describes a fix for `fixed_diagnostic` that may have quick fix
+/// edits available, `noqa` comment edits, or both.
 #[derive(Clone, Debug)]
 pub(crate) struct DiagnosticFix {
     pub(crate) fixed_diagnostic: lsp_types::Diagnostic,
     pub(crate) title: String,
     pub(crate) code: String,
-    pub(crate) edits: Vec<lsp_types::TextEdit>,
+    pub(crate) edits: Option<Vec<lsp_types::TextEdit>>,
```

Comment:
a) I think it could make sense to start documenting the fields. 

b) What's the motivation for using an `Option<Vec<_>>`? Could we just use an empty vec in case there are no edits?

Reply:
> b) What's the motivation for using an Option<Vec<_>>? Could we just use an empty vec in case there are no edits?

There is a distinction in the codebase between an empty vec and a `None` value (a fix with an empty vec would result in a code action that doesn't make any edits). But since 'a code action that doesn't make any edits' shouldn't be sent in the first place, I support your suggestion to make an empty vec represent the lack of a quick fix.